viabel/optimizers.py

```diff
@@ -0,0 +1,808 @@
+import autograd.numpy as np
+import autograd.numpy.random as npr
+import autograd.scipy.stats.multivariate_normal as mvn
+import autograd.scipy.stats.t as t_dist
+from autograd.scipy.linalg import sqrtm
+
+from functools import partial
+import matplotlib.pyplot as plt
+
+import tqdm
+import scipy.stats as stats
+from  .optimization_diagnostics import autocorrelation, monte_carlo_se, monte_carlo_se2, compute_khat_iterates, \
+    gpdfit, montecarlo_se, compute_R_hat, compute_R_hat_halfway, compute_R_hat_halfway_light
+
+
+__all__ = [
+    'adagrad_workflow_optimize',
+    'rmsprop_workflow_optimize',
+    'adam_workflow_optimize'
+]
+
+def learning_rate_schedule(n_iters, learning_rate, learning_rate_end):
+    if learning_rate < 0:
+        raise ValueError('learning rate must be positive')
+    if learning_rate_end is not None:
+        if learning_rate <= learning_rate_end:
+            raise ValueError('initial learning rate must be greater than final learning rate')
+        # constant learning rate for first quarter, then decay like a/(b + i)
+        # for middle half, then constant for last quarter
+        b = n_iters*learning_rate_end/(2*(learning_rate - learning_rate_end))
+        a = learning_rate*b
+        start_decrease_at = n_iters//4
+        end_decrease_at = 3*n_iters//4
+    for i in range(n_iters):
+        if learning_rate_end is None or i < start_decrease_at:
+            yield learning_rate
+        elif i < end_decrease_at:
+            yield a / (b + i - start_decrease_at + 1)
+        else:
+            yield learning_rate_end
+
+def adagrad_workflow_optimize(n_iters, objective_and_grad, init_param,K,
```

Comment:
Please add some additional comments to the optimize functions that explain what each part is doing. Perhaps it's possible to break these functions into smaller pieces  to make the code more self-documenting?